Library/Homebrew/cask/artifact/moved.rb

```diff
@@ -79,7 +79,10 @@ def move(adopt: false, force: false, verbose: false, command: nil, **options)
         if target.dirname.writable?
           FileUtils.move(source, target)
         else
-          command.run!("/bin/mv", args: [source, target], sudo: true)
+          # default sudo user isn't necessarily able to write to Homebrew's locations
+          # e.g. with runas_default set in the sudoers (5) file.
+          command.run!("/bin/cp", args: ["-pR", source, target], sudo: true)
```

Comment:
Hmm, does this only apply here or to pretty much everywhere we use `sudo: true`? I suspect the latter in which case perhaps it's better fixed in the `sudo` logic to ensure that you're `sudo`ing to `root`?

Reply:
With respect to casks, I think, it is proper to act out of assumption that current user has unrestricted access to the homebrew prefix, while sudo has access outside of it. That appears to be sufficient, as sudo is used to move files into appropriate *system* locations. Note that pkgs have their own way to elevate privileges and don't need sudo.

Overall, I think it's important that brew does not unnecessarily limit sysop's ability to configure sudoers.

Reply:
I'm pretty sure the expectation is that `sudo: true` will have unrestricted access to the system (bar those things that SIP prevents).

> Overall, I think it's important that brew does not unnecessarily limit sysop's ability to configure sudoers.

Not sure I agree with this, sorry. I think it's not on Homebrew to support more esoteric default configurations here that will cause issues for many other tools assuming e.g. `sudo ...` will run as `root`.

Reply:
Consider the following argument instead: it is already recommended that in multi-user environments ‘brew’ should have a separated dedicated user. With that in mind ‘mv’ is problematic because you will end up with improperly owned (non-admin user/group) items in system locations, such as /Applications, e.g. see the following thread on Apple’s dev forums: https://developer.apple.com/forums/thread/723134. ‘cp’ addresses that as well my esoteric sudo config :)

Reply:
> it is already recommended that in multi-user environments ‘brew’ should have a separated dedicated user

We don't recommend this anywhere I'm aware of. There's many holes in our multiuser environment setup.

> ‘cp’ addresses that as well my esoteric sudo config :)

Sure, it just feels very weird to have this in `artifact/moved` when it's no longer moving and for it to only be needed with this unusual sudo config.

Reply:
https://docs.brew.sh/FAQ#why-does-homebrew-say-sudo-is-bad

> If you need to run Homebrew in a multi-user environment, consider creating a separate user account specifically for use of Homebrew.

It works quite well for me (with the exception of this problem).

> it just feels very weird to have this in artifact/moved when it's no longer moving and for it to only be needed with this unusual sudo 

In this case consider replacing “move” with “install to a system location” while reasoning.

Move preserves ownership of the file. In single-admin-user environment it does not matter, but with brew running under its own user you it’s a different story

The thread I linked hints that file ownership is important and may lead to silent errors.

Note that this is a non-issue for pkgs as they can elevate their privileges on their own and will install files with proper ownership (and permissions).

Reply:
> If you need to run Homebrew in a multi-user environment, consider creating a separate user account specifically for use of Homebrew.

Sure, I wouldn't consider that "supported" though.

---

Could you make this code conditional on your use case instead of changing it for everyone?

Reply:
I think the current implementation (before my change) is already conditional: most users (single-admin-user running brew) will be able to write to system locations because of the admin group and will hit the `if` branch.

Reply:
@Kentzo Ok, let's try it. Note, we'll revert this ASAP if there's any issues.

Reply:
@Kentzo `rsync` is overkill here. If `cp` isn't enough: let's just revert this.

Reply:
`cp` is good enough for me. It's not good enough for someone who has a separate user for brew without overriding `runas_default` in sudoers.

I do believe that this is a bug in macOS (that it quietly keeps quarantining the app if it's owned by a wrong user/group). Hopefully it will get fixed, but at least I wanted to mention a workaround here if someone else hits it.

Reply:
> It's not good enough for someone who has a separate user for brew without overriding runas_default in sudoers.

Let's not worry about bugs that no users have reported.